glitch/parsers/cmof.py

```diff
@@ -1099,12 +1180,15 @@ def get_code(ce):
                 variable.code = get_code(codeelement)
                 return variable
             else:
-                res = []
+                variable: Variable = Variable(name, None, False)
+                variable.line, variable.column = codeelement.line, codeelement.col
+                variable.code = get_code(codeelement) 
                 for key, value in temp_value.items():
-                    res.append(PuppetParser.__process_codeelement(
-                            puppetmodel.Assignment(codeelement.line, codeelement.col, 
-                                    codeelement.end_line, codeelement.end_col, name + "." + key, value), path, code))
-                return res
+                    variable.variables.append(PuppetParser.__process_codeelement(
+                        puppetmodel.Assignment(codeelement.line, codeelement.col,
+                                               codeelement.end_line, codeelement.end_col, key, value), path, code))
+                
+                return variable
         elif (isinstance(codeelement, puppetmodel.PuppetClass)):
             # FIXME there are components of the class that are not considered
             unit_block: UnitBlock = UnitBlock(
```

Comment:
For both Puppet and Chef, the attributes can not be hierarchical? For instance, for Chef, in variables, you decided that hash values are handled as an hierarchy of variables. Can't the same happen for attributes? Why is that case different?

For instance:
```
file  { '/etc/passwd':
  owner => {
    owner => 'root',
  },
  group => 'root',
}
```
The following Puppet file is parsed as an attribute with an hash value. i do not think that this is wrong, but I would like to understand your point. Thanks!

Reply:
Sorry I think I did not understand your question. The changes i made was regarding only the cases that were being handled with "." . In glitch original code, if the codeelement was a puppetmodel.Assignment, then it would create a variable (not an attribute). Instead of returning a variable with name1.name2.name3, it will return a variable named name1 composed by its child variables (name2 composed by name3).

Reply:
Okay I understand! I think it should be changed if in Puppet and Chef it is also possible to have recursive attributes. Could you give me an example of recursive attributes in Chef? I was searching and in Chef, we can have brace_block and do_block, brace_block corresponds to '{}' and do_block corresponds to 'do end'. But in Chef I think this could be problematic because these two types of blocks can contain any ruby code, and not only definition of new attributes. What do you think?

Reply:
As in Puppet, Ruby also has hashes (https://ruby-doc.org/core-2.5.1/Hash.html) that can be used as values. It is actually the same thing as in the example I showed for Puppet, Honestly, I do not know if it is the right approach to consider these as hierarchical attributes, but I think it is what we are currently (in this PR) doing for variables (please correct me if I am wrong). @jff can you give us your input as well? Thanks!

Reply:
Here is an example of an hash as value of an attribute. It is basically the same as in Puppet.

```
test 'test' do
  test 'a' => 'c', 'b' => 'd'
end
```

The resource [bash](https://docs.chef.io/resources/bash/#examples) also has an attribute of type Hash: `environment`. Other resources can be found with hashes as values.

```
bash 'install_ruby_build' do
  cwd "#{Chef::Config[:file_cache_path]}/ruby-build"
  user 'rbenv'
  group 'rbenv'
  code <<-EOH
    ./install.sh
  EOH
  environment 'PREFIX' => '/usr/local'
end
```

Reply:
Thank you for the examples. The problem is that the Chef parser treats Hash values as normal values (eg: strings), and we cannot differentiate Hash values from the other values (because they are represented by a string). Maybe this is the reason why Glitch does not support hierarchical attributes for Chef.

Reply:
The reason why GLITCH does not support hierarchical attributes for Chef is because when I implemented this I considered that the Hash value is not the same as having an hierarchical attribute since it is not explicit in the syntax (in Ansible the hierarchy is expressed without using a data structure). However, I brought this up because we have to choose if we consider hierarchies with Hash values or not. With one option or the other, I think the behavior should be the same for attributes and variables and also the same for Chef and Puppet. If it is necessary, the parser should be changed to treat hash values differently (it does not change much). I think we should wait for @jff's feedback. I already pinged him on Slack.

Reply:
I think the decision must be based on the smells that will be implemented next. The rationale to support the hierarchical view of variables is based on the catalog of smells that João plans to implement, right? 

On the other hand, if we extend to GLITCH to better support hashes, we will probably get a more precise analysis for the detection of design & implementation smells such as "Improper alignment" (this was a problem identified in our preliminary study / tool paper).

We can focus on variables in this PR and, depending on João's project timeline, we can consider hashes in a separate PR.

Reply:
Ok that makes sense @jff. However, what I was trying to ask was if you consider it makes sense to consider hashes as hierarchical variables in Puppet and Chef (it is the current behavior). For instance,

```
test = {
	a => 1,
	b => 2
}
```

is being parsed as a variable that has two variables 'a' and 'b'.  I think in previous versions there was inconsistencies in the handling of hashes as well, but I am trying to understand what is the best path to follow here. The options:

1. In "regular" assignments, hashes are handled as hierarchical variables. Inside atomic units, in attributes, hashes are considered values (I think this is how it is currently implemented @joaotgoncalves please confirm).
2. Hashes are handled as hierarchical variables and attributes.
3. Hashes are always considered values.

The difference between Puppet/Chef and Ansible is that in Ansible the attributes and variables are defined as hierarchical by the syntax of YAML:
```
test:
    a: 1
    b: 2
```
In this case, I think it is not dubious. I think the same happens for Terraform.

Reply:
1. Yes, in the current version of GLITCH, regular assignments hashes are handled as hierarchical variables. For attributes is only for Ansible (just like @Nfsaavedra explained).
2. In Terraform, it happens the same as it happens with Ansible: attributes and variables are defined as hierarchical by the syntax of HCL.
3. I agree with both of you (we have to decide if in Puppet and Chef, hash values are considered as hierarchical attributes as it happens with Ansible, but this might delay my work, so I prefer to move on with the implementation of Terraform as it is more important right now). 
4. I am going to finish the modifications soon (between today and tomorrow), and push the new code related to this PR.

Reply:
Note to myself (@Nfsaavedra): I'm finally deciding that in Puppet and Chef, hash values will not be considered as hierarchical variables, since now we are introducing hashes as part of the IR.